Mathlib/Data/Real/IsNonarchimedean.lean

```diff
@@ -0,0 +1,152 @@
+/-
+Copyright (c) 2024 María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos-Fernández
+-/
+import Mathlib.Algebra.Module.Defs
+import Mathlib.Data.Nat.Choose.Sum
+import Mathlib.Data.Real.Basic
+
+/-!
+# Nonarchimedean functions
+
+A function `f : R → ℝ≥0` is nonarchimedean if it satisfies the strong triangle inequality
+`f (r + s) ≤ max (f r) (f s)` for all `r s : R`. This file proves basic properties of
+nonarchimedean functions.
+-/
+
+/-- A function `f : R → ℝ≥0` is nonarchimedean if it satisfies the strong triangle inequality
+  `f (r + s) ≤ max (f r) (f s)` for all `r s : R`. -/
+def IsNonarchimedean {A : Type*} [Add A] (f : A → ℝ) : Prop :=
+  ∀ r s, f (r + s) ≤ max (f r) (f s)
+
+namespace IsNonarchimedean
+
+/-- A nonarchimedean function satisfies the triangle inequality. -/
+theorem add_le {α : Type*} [AddCommGroup α] {f : α → ℝ} (hf : ∀ x : α, 0 ≤ f x)
+    (hna : IsNonarchimedean f) (a b : α) : f (a + b) ≤ f a + f b := by
```

Comment:
Is there any specific reason why you want `a` and `b`  explicit?

Reply:
Not really; I have made them implicit.